### DISCUSSION TITLE
Return back BeginSend/EndSend to HttpClient for best performance

### DISCUSSION THREAD

**Discussion by okarpov** [dotnet/runtime#86867] [Ideas] [answered]
As Microsoft mentiones Begin/End pattern more performant than Async/await

So i do suggest to return back Begin/End putterns especially for Send methods to get best performance in applications which do extensive sends.

**Comment by stephentoub** [comment]
> As Microsoft mentiones Begin/End pattern more performant than Async/await

Where does "Microsoft mention" that? It is inaccurate and we're not going to add new such begin/end methods.

**Reply by okarpov** [reply]
Doesn't async/await adds extra layer as Task ?

**Reply by stephentoub** [reply]
No, the Task APIs in .NET Core are not layered on top of Begin/End, at least not anywhere it matters. In most cases it's the opposite.

**Reply by okarpov** [reply]
https://learn.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/deciding-when-to-implement-the-event-based-asynchronous-pattern?redirectedfrom=MSDN

**Reply by stephentoub** [reply]
Those docs are from circa 2005. They should be obsoleted or deleted.
cc: @gewarren

**Reply by okarpov** [reply]
> it says "Article 09/15/2021 15 contributors"°

That just means some edit was made to it on that date, eg fixing a typo.

**Reply by stephentoub** [reply]
They also say nothing about tasks...the "event-based pattern" discussed there is not about Task nor async/await.

If you haven't already, I'd encourage you to read https://devblogs.microsoft.com/dotnet/how-async-await-really-works/.

**Reply by okarpov** [reply]
> No, the Task APIs in .NET Core are not layered on top of Begin/End, at least not anywhere it matters. In most cases it's the opposite.

sorry, i did not mean its layered on top of Begin/End, i meant in general it adds extra layer as Task+Pool which consumes time and memory

i'm talking about it seeing how much memory and cpu  is being cosumed by Strings and async/await sending +1000 QPS using SendAsync(System.Net.Http.HttpRequestMessage)

**Reply by stephentoub** [reply]
Whatever consumption you're seeing with tasks and async/await, it will be higher with begin/end.

**Reply by okarpov** [reply]
hmm, interesting.

**Reply by okarpov** [reply]
and what is about performance?

**Reply by stephentoub** [reply]
> and what is about performance?

I don't understand the question. Can you elaborate?

**Reply by okarpov** [reply]
so Microsoft mentioned that Begin/End has performance benefits, but you said it is obsolet and wrold has changed.
You confirmed that RAM consuming is less with async/await pattern than the any other pattern.
So my final question is: does performance of the asyn/await also higher than any other pattern?

**Reply by stephentoub** [reply]
The {Value}Task-based APIs along with async/await offer the best performance and usability in general.

**Reply by okarpov** [reply]
you mean ```Task<value>``` only but not ```Task<object>```?

**Reply by stephentoub** [reply, marked as answer]
No, I mean APIs returning `Task`, `Task<T>`, `ValueTask`, and `ValueTask<T>`.

Again, I'd recommend you read the article I linked.
